File: app/src/main/java/com/druk/servicebrowser/ui/viewmodel/RegTypeBrowserViewModel.java
package com.druk.servicebrowser.ui.viewmodel;

import android.app.Application;
import android.net.nsd.NsdManager;
import android.net.nsd.NsdServiceInfo;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.druk.servicebrowser.BonjourApplication;
import com.druk.servicebrowser.BonjourServiceInfo;
import com.druk.servicebrowser.Config;
import com.druk.servicebrowser.ServiceTypeResolver;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Discovers available mDNS service types on the local network.
 * <p>
 * NsdManager does not support the _services._dns-sd._udp meta-query, so we perform
 * it via a persistent mDNS multicast listener ({@link ServiceTypeResolver}). Each new
 * service type is immediately handed to NsdManager for per-service browsing.
 */
public class RegTypeBrowserViewModel extends AndroidViewModel {

    private static final String TAG = "RegTypeBrowserVM";

    /** Delay before retrying types that failed with FAILURE_MAX_LIMIT. */
    private static final long RETRY_DELAY_MS = 3000;

    private final NsdManager nsdManager;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final ServiceTypeResolver resolver = new ServiceTypeResolver();

    /** Active NsdManager discovery listeners keyed by service type. */
    private final HashMap<String, NsdManager.DiscoveryListener> activeListeners = new HashMap<>();

    /** Types that found at least one service. Thread-safe. */
    private final ConcurrentHashMap<String, BonjourDomain> foundTypes = new ConcurrentHashMap<>();

    /** Queue of types that hit FAILURE_MAX_LIMIT and need retry. */
    private final Queue<String> retryQueue = new LinkedList<>();

    private final MutableLiveData<Collection<BonjourDomain>> servicesLiveData = new MutableLiveData<>();
    private final MutableLiveData<Throwable> errorLiveData = new MutableLiveData<>();

    private final Runnable retryRunnable = () -> {
        synchronized (RegTypeBrowserViewModel.this) {
            drainQueue(retryQueue);
        }
    };

    public RegTypeBrowserViewModel(@NonNull Application application) {
        super(application);
        nsdManager = BonjourApplication.getNsdManager(application);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        resolver.stop();
        handler.removeCallbacksAndMessages(null);
        executor.shutdownNow();
        synchronized (this) {
            for (NsdManager.DiscoveryListener listener : activeListeners.values()) {
                try {
                    nsdManager.stopServiceDiscovery(listener);
                } catch (IllegalArgumentException ignored) {
                }
            }
            activeListeners.clear();
        }
        foundTypes.clear();
    }

    public ConcurrentHashMap<String, BonjourDomain> getServices() {
        return foundTypes;
    }

    public LiveData<Collection<BonjourDomain>> getServicesLiveData() {
        return servicesLiveData;
    }

    public LiveData<Throwable> getErrorLiveData() {
        return errorLiveData;
    }

    public void startDiscovery() {
        // Run the mDNS listener on a background thread — it stays open until onCleared
        executor.execute(() -> resolver.start(serviceType -> {
            // Called on the background thread for each newly discovered type — start browsing immediately
            handler.post(() -> {
                synchronized (RegTypeBrowserViewModel.this) {
                    if (!activeListeners.containsKey(serviceType)) {
                        startSingleDiscovery(serviceType);
                    }
                }
            });
        }));
    }

    /** Try to start all queued types; stop on first FAILURE_MAX_LIMIT. */
    private void drainQueue(Queue<String> queue) {
        while (!queue.isEmpty()) {
            String type = queue.poll();
            if (type == null || activeListeners.containsKey(type)) continue;
            if (!startSingleDiscovery(type)) {
                queue.add(type); // put it back
                handler.postDelayed(retryRunnable, RETRY_DELAY_MS);
                break;
            }
        }
    }

    /**
     * @return true if started successfully, false if hit system limit
     */
    private synchronized boolean startSingleDiscovery(String serviceType) {
        NsdManager.DiscoveryListener listener = new NsdManager.DiscoveryListener() {
            @Override
            public void onDiscoveryStarted(String regType) {
                Log.d(TAG, "Browsing: " + regType);
            }

            @Override
            public void onServiceFound(NsdServiceInfo nsdServiceInfo) {
                handleServiceEvent(serviceType, false);
            }

            @Override
            public void onServiceLost(NsdServiceInfo nsdServiceInfo) {
                handleServiceEvent(serviceType, true);
            }

            @Override
            public void onDiscoveryStopped(String regType) {
            }

            @Override
            public void onStartDiscoveryFailed(String regType, int errorCode) {
                Log.w(TAG, "Browse failed for " + regType + ": " + errorCode);
                synchronized (RegTypeBrowserViewModel.this) {
                    activeListeners.remove(serviceType);
                }
                if (errorCode == NsdManager.FAILURE_MAX_LIMIT) {
                    synchronized (RegTypeBrowserViewModel.this) {
                        retryQueue.add(serviceType);
                    }
                    handler.removeCallbacks(retryRunnable);
                    handler.postDelayed(retryRunnable, RETRY_DELAY_MS);
                }
            }

            @Override
            public void onStopDiscoveryFailed(String regType, int errorCode) {
                Log.w(TAG, "Stop failed for " + regType + ": " + errorCode);
            }
        };

        activeListeners.put(serviceType, listener);

        try {
            nsdManager.discoverServices(serviceType, NsdManager.PROTOCOL_DNS_SD, listener);
            return true;
        } catch (IllegalArgumentException e) {
            Log.w(TAG, "Failed to start browse for " + serviceType, e);
            activeListeners.remove(serviceType);
            return false;
        }
    }

    private void handleServiceEvent(String serviceType, boolean lost) {
        String[] parts = serviceType.split("\\.");
        if (parts.length < 2) return;

        String serviceName = parts[0];
        String protocolSuffix = parts[1];

        BonjourDomain domain = foundTypes.get(serviceType);
        if (domain == null) {
            BonjourServiceInfo info = new BonjourServiceInfo.Builder()
                    .serviceName(serviceName)
                    .regType(protocolSuffix + "." + Config.LOCAL_DOMAIN)
                    .domain(Config.EMPTY_DOMAIN)
                    .build();
            domain = new BonjourDomain(info);
            foundTypes.put(serviceType, domain);
        }

        if (lost) {
            domain.serviceCount = Math.max(0, domain.serviceCount - 1);
        } else {
            domain.serviceCount++;
        }

        servicesLiveData.postValue(new ArrayList<>(foundTypes.values()));
    }

    public static class BonjourDomain extends BonjourServiceInfo {
        public int serviceCount = 0;

        public BonjourDomain(BonjourServiceInfo info) {
            super(new Builder(info));
        }
    }
}

File: app/src/main/java/com/druk/servicebrowser/ui/viewmodel/RegTypeBrowserViewModel.kt
package com.druk.servicebrowser.ui.viewmodel

import android.app.Application
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.druk.servicebrowser.BonjourApplication
import com.druk.servicebrowser.BonjourServiceInfo
import com.druk.servicebrowser.Config
import com.druk.servicebrowser.ServiceTypeResolver
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import android.os.Handler
import android.os.Looper

data class BonjourDomain(
    val info: BonjourServiceInfo,
    var serviceCount: Int = 0
) {
    // Delegate identity to the underlying BonjourServiceInfo
    val serviceName: String get() = info.displayName
    val regType: String? get() = info.regType
    val domain: String? get() = info.domain

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BonjourDomain) return false
        return info == other.info
    }

    override fun hashCode(): Int = info.hashCode()
}

class RegTypeBrowserViewModel(application: Application) : AndroidViewModel(application) {

    private val nsdManager = BonjourApplication.getNsdManager(application)
    private val resolver = ServiceTypeResolver()

    private val handler = Handler(Looper.getMainLooper())
    private val activeListeners = HashMap<String, NsdManager.DiscoveryListener>()
    private val foundTypes = HashMap<String, BonjourDomain>()
    private val retryChannel = Channel<String>(Channel.UNLIMITED)

    private val _services = MutableStateFlow<List<BonjourDomain>>(emptyList())
    val services: StateFlow<List<BonjourDomain>> = _services

    private val _error = MutableSharedFlow<Throwable>(extraBufferCapacity = 1)
    val error: SharedFlow<Throwable> = _error

    // LiveData-style accessors for fragment compatibility
    fun getServicesLiveData() = _services
    fun getErrorLiveData() = _error
    fun getServices() = foundTypes

    override fun onCleared() {
        super.onCleared()
        handler.removeCallbacksAndMessages(null)
        for (listener in activeListeners.values) {
            try {
                nsdManager.stopServiceDiscovery(listener)
            } catch (_: IllegalArgumentException) {
            }
        }
        activeListeners.clear()
        foundTypes.clear()
        retryChannel.close()
    }

    fun startDiscovery() {
        // Collect service types from the resolver
        viewModelScope.launch(Dispatchers.IO) {
            resolver.serviceTypes().collect { serviceType ->
                launch(Dispatchers.Main) {
                    if (serviceType !in activeListeners) {
                        startSingleDiscovery(serviceType)
                    }
                }
            }
        }

        // Process retry queue
        viewModelScope.launch(Dispatchers.Main) {
            for (type in retryChannel) {
                if (type in activeListeners) continue
                if (!startSingleDiscovery(type)) {
                    delay(RETRY_DELAY_MS)
                    retryChannel.trySend(type)
                }
            }
        }
    }

    private fun startSingleDiscovery(serviceType: String): Boolean {
        val listener = object : NsdManager.DiscoveryListener {
            override fun onDiscoveryStarted(regType: String) {
                Log.d(TAG, "Browsing: $regType")
            }

            override fun onServiceFound(nsdServiceInfo: NsdServiceInfo) {
                handleServiceEvent(serviceType, false)
            }

            override fun onServiceLost(nsdServiceInfo: NsdServiceInfo) {
                handleServiceEvent(serviceType, true)
            }

            override fun onDiscoveryStopped(regType: String) {}

            override fun onStartDiscoveryFailed(regType: String, errorCode: Int) {
                Log.w(TAG, "Browse failed for $regType: $errorCode")
                handler.post {
                    activeListeners.remove(serviceType)
                }
                if (errorCode == NsdManager.FAILURE_MAX_LIMIT) {
                    retryChannel.trySend(serviceType)
                }
            }

            override fun onStopDiscoveryFailed(regType: String, errorCode: Int) {
                Log.w(TAG, "Stop failed for $regType: $errorCode")
            }
        }

        activeListeners[serviceType] = listener

        return try {
            nsdManager.discoverServices(serviceType, NsdManager.PROTOCOL_DNS_SD, listener)
            true
        } catch (e: IllegalArgumentException) {
            Log.w(TAG, "Failed to start browse for $serviceType", e)
            activeListeners.remove(serviceType)
            false
        }
    }

    private fun handleServiceEvent(serviceType: String, lost: Boolean) {
        handler.post {
            val parts = serviceType.split(".")
            if (parts.size < 2) return@post

            val serviceName = parts[0]
            val protocolSuffix = parts[1]

            val domain = foundTypes.getOrPut(serviceType) {
                val info = BonjourServiceInfo(
                    serviceName = serviceName,
                    regType = "$protocolSuffix.${Config.LOCAL_DOMAIN}",
                    domain = Config.EMPTY_DOMAIN
                )
                BonjourDomain(info)
            }

            if (lost) {
                domain.serviceCount = maxOf(0, domain.serviceCount - 1)
            } else {
                domain.serviceCount++
            }

            _services.value = ArrayList(foundTypes.values)
        }
    }

    companion object {
        private const val TAG = "RegTypeBrowserVM"
        private const val RETRY_DELAY_MS = 3000L
    }
}
